### PR TITLE
Fix tab events and test event nested event subscriptions

### PR DESCRIPTION
**Pull request by ericvicenti**
Fixes tabs bug mentioned by @noxee in #3345

Also adds a test for the event subscription implementation

### CODE REVIEW

File: src/getChildEventSubscriber.js [hunk 1]
```diff
@@ -46,13 +32,12 @@ export default function getChildEventSubscriber(
...
     subscribers &&
       subscribers.forEach(subs => {
-        // $FlowFixMe - Payload should probably understand generic state type
         subs(payload);
       });
   };
 
-  let isSelfFocused = false;
-
+  let isParentFocused = true;
```
Comment (brentvatne) [inline src/getChildEventSubscriber.js]: can we clarify in a comment somewhere here what parent and child mean in this context, and why parent initializes as focused and child as unfocused?

**Comment by ericvicenti** [comment]
Ok I'm ready for a real review of this. The core logic has been re-implemented with a fraction of the LOC, @skevy's bug has been tested, and I actually understand the implementation this time

File: src/getChildEventSubscriber.js [hunk 2]
```diff
@@ -55,110 +59,58 @@ export default function getChildEventSubscriber(addListener, key) {
...
+        } else if (eventName === 'action' && isFocused) {
+          emit((lastEmittedEvent = 'willFocus'), childPayload);
         }
-        if (isFocused && !isTransitioning && eventName === 'didBlur') {
-          emit(childPayload);
+      }
+      if (lastEmittedEvent === 'willFocus') {
+        if (eventName === 'didFocus' && isFocused && !isTransitioning) {
+          emit((lastEmittedEvent = 'didFocus'), childPayload);
+        } else if (!isTransitioning && isFocused && eventName === 'action') {
```
Comment (brentvatne) [inline src/getChildEventSubscriber.js]: let's keep the same order for the conditionals here, 

`if (eventName === 'didFocus' && isFocused && !isTransitioning)` ✅ 
`else if (!isTransitioning && isFocused && eventName === 'action')` 🚫
`else if (eventName === 'action' && isFocused && !isTransitioning)` ✅

Comment (brentvatne) [reply src/getChildEventSubscriber.js]: (same applies for other conditionals around here)

File: src/__tests__/getChildEventSubscriber-test.js [hunk 3]
```diff
@@ -213,67 +213,102 @@ test('grandchildren transitions', () => {
   });
   expect(childWillBlurHandler.mock.calls.length).toBe(1);
   expect(childDidBlurHandler.mock.calls.length).toBe(1);
-  expect(childActionHandler.mock.calls.length).toBe(3);
+  expect(childActionHandler.mock.calls.length).toBe(1);
 });
 
-test('pass through focus', () => {
```
[the remark this answers is not in the document]
Comment (ericvicenti) [reply src/__tests__/getChildEventSubscriber-test.js]: its a matter of opinion I guess. Personally I like that these perform the test directly. What API would you prefer to test with?

File: src/getChildEventSubscriber.js [hunk 4]
```diff
@@ -28,16 +28,20 @@ export default function getChildEventSubscriber(addListener, key) {
...
     subscribers &&
       subscribers.forEach(subs => {
-        subs(payload);
+        subs(payloadWithType);
       });
   };
 
   let isParentFocused = true;
-  let isChildFocused = false;
+  let hasEmittedChildFocusChange = true;
```
[the remark this answers is not in the document]
Comment (ericvicenti) [reply src/getChildEventSubscriber.js]: willfix

File: src/__tests__/getChildEventSubscriber-test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (brentvatne) [reply src/__tests__/getChildEventSubscriber-test.js]: ok i don't feel too strongly about it. i just think that we could probably refactor the setup into a function that sits outside of the actual test block so we can focus on the assertions in the test but perhaps it wouldn't help too much here if we can't reuse the setup across multiple tests

**Pull request merged**